flink-runtime/src/main/java/org/apache/flink/runtime/rpc/MainThreadValidatorUtil.java

```diff
@@ -44,4 +51,24 @@ public void exitMainThread() {
 		assert(endpoint.currentMainThread.compareAndSet(Thread.currentThread(), null)) :
 				"The RpcEndpoint has concurrent access from " + endpoint.currentMainThread.get();
 	}
+
+	/**
+	 * Returns true iff the current thread is equals to the provided expected thread and logs violations.
+	 *
+	 * @param expected the expected main thread.
+	 * @return true iff the current thread is equals to the provided expected thread.
+	 */
+	public static boolean isRunningInExpectedThread(@Nullable Thread expected) {
+		Thread actual = Thread.currentThread();
+		if (expected != actual) {
+			log.warn("Violation of main thread constraint detected: expected <{}> but running in <{}>.",
+				expected,
+				actual,
+				new Exception());
```

Comment:
Did you create the `Exception` here to capture the stack trace?

Reply:
We could add the logging message to the exception as well.

Reply:
Yes, and yes ok.